test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -1052,21 +1053,38 @@ private synchronized void reset(boolean wipeData) throws IOException {
         logger.debug("Cluster is consistent again - nodes: [{}] nextNodeId: [{}] numSharedNodes: [{}]", nodes.keySet(), nextNodeId.get(), newSize);
     }
 
-    /** ensure a cluster is form with {@link #nodes}.size() nodes. */
+    /** ensure a cluster is formed with all published nodes. */
     private void validateClusterFormed() {
         String name = randomFrom(random, getNodeNames());
         validateClusterFormed(name);
     }
 
-    /** ensure a cluster is form with {@link #nodes}.size() nodes, but do so by using the client of the specified node */
+    /** ensure a cluster is formed with all published nodes, but do so by using the client of the specified node */
     private void validateClusterFormed(String viaNode) {
-        final int size = nodes.size();
-        logger.trace("validating cluster formed via [{}], expecting [{}]", viaNode, size);
+        Set<DiscoveryNode> expectedNodes = new HashSet<>();
```

Comment:
It was actually required for the tests to succeed, as `InternalTestCluster.updateMinMasterNodes` would otherwise be too trigger-happy. One example I saw (`InternalTestClusterTests.testTwoNodeCluster`, rolling restart of two master-eligible nodes) was that it would update the `minimum_master_nodes` from 1 to 2 just in-between when the previous node instance was booted from the cluster and the restarted node rejoined the cluster.